platforms/core-configuration/base-diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/TaskReportTask.java

```diff
@@ -245,44 +252,63 @@ private List<RuleDetails> ruleDetailsFor(Project project) {
 
     private DefaultGroupTaskReportModel taskReportModelFor(Project project, boolean detail) {
         final AggregateMultiProjectTaskReportModel aggregateModel = new AggregateMultiProjectTaskReportModel(!detail, detail, getDisplayGroup(), getDisplayGroups());
-        final TaskDetailsFactory taskDetailsFactory = new TaskDetailsFactory(project);
 
-        final SingleProjectTaskReportModel projectTaskModel = buildTaskReportModelFor(taskDetailsFactory, project);
-        aggregateModel.add(projectTaskModel);
+        ProjectIdentity relativeProjectIdentity = ((ProjectInternal) project).getProjectIdentity();
 
-        for (final Project subproject : project.getSubprojects()) {
-            aggregateModel.add(buildTaskReportModelFor(taskDetailsFactory, subproject));
-        }
+        Stream.concat(Stream.of(project), project.getSubprojects().stream())
+            .map(p -> buildTaskReportModelFor(relativeProjectIdentity, p))
+            .forEach(aggregateModel::add);
 
         aggregateModel.build();
 
         return DefaultGroupTaskReportModel.of(aggregateModel);
     }
 
-    private SingleProjectTaskReportModel buildTaskReportModelFor(
-        final TaskDetailsFactory taskDetailsFactory,
-        final Project subproject
-    ) {
-        return projectStateFor(subproject).fromMutableState(
-            project -> SingleProjectTaskReportModel.forTasks(
-                getProjectTaskLister().listProjectTasks(project),
-                taskDetailsFactory
-            )
-        );
+    private SingleProjectTaskReportModel buildTaskReportModelFor(ProjectIdentity relativeProjectIdentity, Project p) {
+        return ((ProjectInternal) p).getOwner().fromMutableState(project -> {
+            ImmutableSetMultimap.Builder<String, TaskDetails> groups = ImmutableSetMultimap.<String, TaskDetails>builder()
+                .orderKeysBy(String::compareToIgnoreCase)
+                .orderValuesBy(Comparator.comparing(TaskDetails::getPath));
+
+            for (Task task : getProjectTaskLister().listProjectTasks(project)) {
```

Comment:
💭 This is not IP-compatible. I think `ProjectTaskLister` as a concept is not compatible. We should make sure we at least issue IP-violations here and ensure the target project is configured before we ask it for tasks. I wonder if this ever worked with configure-on-demand?

Can you create a follow-up issue for this?

Reply:
This is called within `((ProjectInternal) p).getOwner().fromMutableState(project -> {`

Which is _technically_ safe in regards to project locks. 

This is linked in with the `Cached` field as well, which allows us to also be non-CC compatible here.

Happy to write up an issue on this though, cause there are definitely some odd things going on here.

Reply:
> Which is technically safe in regards to project locks.

Quickly scanning the sources, I don't think this codepath calls `ProjectState.ensureConfigured()`, but maybe I missed it.

Reply:
Hm. I don't think it is then.

I'll include this as part of the issue

Reply:
Added this to the board for now https://github.com/orgs/gradle/projects/148/views/2?pane=issue&itemId=120517195

Reply:
https://github.com/gradle/gradle/issues/34345